pkg/filters/proxy/loadbalance.go

```diff
@@ -147,6 +200,73 @@ func (blb *BaseLoadBalancer) ChooseServer(req *httpprot.Request) *Server {
 	return nil
 }
 
+// chooseServerByServerMatch chooses server by matching server
+func (blb *BaseLoadBalancer) chooseServerByServerMatch(req *httpprot.Request) *Server {
+	cookie, err := req.Cookie(blb.spec.StickySession.LBCookieName)
+	if err != nil {
+		return nil
+	}
+
+	for _, s := range blb.Servers {
+		if match(cookie.Value, []byte(s.ID())) {
+			return s
+		}
+	}
+
+	return nil
+}
+
+// ReturnServer does some custom work before return server
+func (blb *BaseLoadBalancer) ReturnServer(server *Server, req *httpprot.Request, resp *httpprot.Response) {
+	if blb.spec.StickySession == nil {
+		return
+	}
+
+	switch blb.spec.StickySession.Mode {
+	case StickySessionModeCookieServerMatch:
+		cookie := &http.Cookie{
+			Name:    blb.spec.StickySession.LBCookieName,
+			Value:   sign([]byte(server.ID())),
+			Expires: time.Now().Add(time.Duration(blb.spec.StickySession.LBCookieExpire * 1000000)),
+		}
+		resp.SetCookie(cookie)
+	}
+}
+
+// sign signs plain text byte array to encoded string
+func sign(plain []byte) string {
+	signed := make([]byte, KeyLen+sha256.Size)
+	key := signed[:KeyLen]
+	macBytes := signed[KeyLen:]
+
+	// use maphash to generate random key fast
+	binary.LittleEndian.PutUint64(key, new(maphash.Hash).Sum64())
+	mac := hmac.New(sha256.New, key)
+	mac.Write(plain)
+	mac.Sum(macBytes[:0])
+
+	return hex.EncodeToString(signed)
+}
+
+// match matches plain text byte array for encoded string
+func match(encoded string, plain []byte) bool {
+	signed, err := hex.DecodeString(encoded)
+	if err != nil {
+		return false
+	}
+
+	if len(signed) != KeyLen+sha256.Size {
+		return false
+	}
```

Comment:
we can do this in `chooseServerByServerMatch` to improve performance.
and we can also move the rest of this function into `chooseServerByServerMatch` as it is very simple and has no other usages.

Reply:
good suggestion, applied